umh-core/pkg/fsm/dataflowcomponent/reconcile.go

```diff
@@ -32,6 +40,332 @@ import (
+		err := d.baseFSMInstance.GetBackoffError(snapshot.Tick)
+		d.baseFSMInstance.GetLogger().Debugf("Skipping reconcile for Dataflowcomponent pipeline %s: %v", dataflowComponentInstanceName, err)
+
+		// if it is a permanent error, start the removal process and reset the error (so that we can reconcile towards a stopped / removed state)
+		if backoff.IsPermanentFailureError(err) {
+
+			// if it is already in stopped, stopping, removing states, and it again returns a permanent error,
+			// we need to throw it to the manager as the instance itself here cannot fix it anymore
+			if d.IsRemoved() || d.IsRemoving() || d.IsStopping() || d.IsStopped() {
+				d.baseFSMInstance.GetLogger().Errorf("DataflowComponent instance %s is already in a terminal state, force removing it", dataflowComponentInstanceName)
+				// force delete everything from the s6 file directory
+				d.service.ForceRemoveDataFlowComponent(ctx, filesystemService, dataflowComponentInstanceName)
+				return err, false
+			} else {
+				d.baseFSMInstance.GetLogger().Errorf("DataflowComponent instance %s is not in a terminal state, resetting state and removing it", dataflowComponentInstanceName)
+				d.baseFSMInstance.ResetState()
+				d.Remove(ctx)
+				return nil, false // let's try to at least reconcile towards a stopped / removed state
+			}
+		}
+		return nil, false
+	}
+
+	// Step 2: Detect external changes.
+	err = d.reconcileExternalChanges(ctx, filesystemService, snapshot.Tick)
+	switch {
+	case err == nil:
+	// All good. Do nothing and continue to reconcile
+	case errors.Is(err, dataflowcomponentservice.ErrServiceNotExists):
+		// If service doesn't exists, it will be created in the next reconcile loop. So set the error to nil
+		err = nil
+	case errors.Is(err, context.DeadlineExceeded):
+		// Healthchecks occasionally take longer (sometimes up to 70ms),
+		// resulting in context.DeadlineExceeded errors. In this case, we want to
+		// mark the reconciliation as complete for this tick since we've likely
+		// already consumed significant time. We return reconciled=true to prevent
+		// further reconciliation attempts in the current tick.
+		return nil, true // We don't want to return an error here, as this can happen in normal operations
+	default:
+		// Consider a special case for DFC FSM here
+		// While creating for the first time, reconcileExternalChanges function will throw an error such as
+		// s6 config file not found in the path since DFC fsm is relying on BenthosFSM and Benthos in turn relies on S6 fsm
+		// Inorder for DFC fsm to start, benthosManager.Reconcile should be called and this is called at the end of the function
+		// So set the err to nil in this case
+		// An example error: "failed to update observed state: failed to get observed DataflowComponent config: failed to get benthos config: failed to get benthos config file for service benthos-dataflow-hello-world-dfc: service does not exist"
+		if strings.Contains(err.Error(), "service does not exist") {
```

Comment:
this is the same as `errors.Is(err, dataflowcomponentservice.ErrServiceNotExists):`

if you need this, then the errors are probably not properly wrapped. there is the option to wrap it.

Reply:
Fixed